File: typescript-nodejs/tests/integration/repository-scenarios.test.ts
import { InMemoryOrderRepository } from '../../examples/infrastructure/repositories/in-memory-order-repository';
import { Order, OrderId, CustomerId } from '../../examples/domain/entities/order';
import { OrderItem } from '../../examples/domain/entities/order-item';
import { Money } from '../../examples/domain/value-objects/money';
import { OrderStatus } from '../../examples/domain/value-objects/order-status';
import { Result, Error } from '../../src/functional';

/**
 * Integration tests for Repository async patterns and operations
 * Tests complete repository workflows including CRUD operations, querying, and error handling
 */
describe('Repository Async Patterns Integration Tests', () => {
    let repository: InMemoryOrderRepository;

    beforeEach(() => {
        repository = new InMemoryOrderRepository();
    });

    afterEach(() => {
        repository.clear();
    });

    describe('Should_HandleBasicCRUDOperations_When_ManagingAggregates', () => {
        it('should add and retrieve order successfully', async () => {
            // Given
            const customerId = new CustomerId('customer-123');
            const order = Order.create(customerId);
            const item = OrderItem.create('Product A', new Money(10.99, 'USD'), 2);
            order.addItem(item);

            // When - Add order
            const addResult = await repository.addAsync(order);

            // Then - Add succeeded
            expect(addResult.isSuccess).toBe(true);

            // When - Retrieve order
            const getResult = await repository.getByIdAsync(order.id);

            // Then - Retrieve succeeded
            expect(getResult.isSuccess).toBe(true);
            expect(getResult.value.hasValue).toBe(true);

            const retrievedOrder = getResult.value.value;
            expect(retrievedOrder.id.equals(order.id)).toBe(true);
            expect(retrievedOrder.customerId.equals(customerId)).toBe(true);
            expect(retrievedOrder.itemCount).toBe(1);
            expect(retrievedOrder.totalAmount.amount).toBe(21.98);
        });

        it('should update order with optimistic concurrency control', async () => {
            // Given
            const customerId = new CustomerId('customer-123');
            const order = Order.create(customerId);
            const item = OrderItem.create('Product A', new Money(10.99, 'USD'), 1);
            order.addItem(item);

            await repository.addAsync(order);
            const initialVersion = order.version;

            // When - Update order
            order.confirm();
            const updateResult = await repository.updateAsync(order);

            // Then - Update succeeded
            expect(updateResult.isSuccess).toBe(true);
            expect(order.version).toBe(initialVersion + 1);

            // When - Retrieve updated order
            const getResult = await repository.getByIdAsync(order.id);
            const retrievedOrder = getResult.value.value;

            // Then - Changes persisted
            expect(retrievedOrder.status).toBe(OrderStatus.CONFIRMED);
            expect(retrievedOrder.version).toBe(order.version);
        });

        it('should handle optimistic concurrency conflicts', async () => {
            // Given
            const customerId = new CustomerId('customer-123');
            const order = Order.create(customerId);
            const item = OrderItem.create('Product A', new Money(10.99, 'USD'), 1);
            order.addItem(item);

            await repository.addAsync(order);

            // When - Simulate concurrent modification
            const order1 = (await repository.getByIdAsync(order.id)).value.value;
            const order2 = (await repository.getByIdAsync(order.id)).value.value;

            order1.confirm();
            await repository.updateAsync(order1); // First update succeeds

            order2.confirm();
            const conflictResult = await repository.updateAsync(order2); // Second update should fail

            // Then - Concurrency conflict detected
            expect(conflictResult.isFailure).toBe(true);
            expect(conflictResult.error.category).toBe('Concurrency');
            expect(conflictResult.error.code).toBe('Repository.ConcurrencyConflict');
        });

        it('should delete order successfully', async () => {
            // Given
            const customerId = new CustomerId('customer-123');
            const order = Order.create(customerId);
            await repository.addAsync(order);

            // When - Delete order
            const deleteResult = await repository.deleteAsync(order.id);

            // Then - Delete succeeded
            expect(deleteResult.isSuccess).toBe(true);

            // When - Try to retrieve deleted order
            const getResult = await repository.getByIdAsync(order.id);

            // Then - Order not found
            expect(getResult.isSuccess).toBe(true);
            expect(getResult.value.hasValue).toBe(false);
        });

        it('should check if order exists', async () => {
            // Given
            const customerId = new CustomerId('customer-123');
            const order = Order.create(customerId);

            // When - Check before adding
            const existsBeforeResult = await repository.existsAsync(order.id);

            // Then - Should not exist
            expect(existsBeforeResult.isSuccess).toBe(true);
            expect(existsBeforeResult.value).toBe(false);

            // When - Add order and check again
            await repository.addAsync(order);
            const existsAfterResult = await repository.existsAsync(order.id);

            // Then - Should exist
            expect(existsAfterResult.isSuccess).toBe(true);
            expect(existsAfterResult.value).toBe(true);
        });
    });

    describe('Should_HandleQueryOperations_When_SearchingOrders', () => {
        beforeEach(async () => {
            // Set up test data
            const customers = ['customer-1', 'customer-2', 'customer-3'];
            const statuses = [OrderStatus.PENDING, OrderStatus.CONFIRMED, OrderStatus.SHIPPED];

            for (let i = 0; i < 10; i++) {
                const customerId = new CustomerId(customers[i % customers.length]!);
                const order = Order.create(customerId);

                const item = OrderItem.create(
                    `Product ${i}`,
                    new Money(10 + i, 'USD'),
                    1 + (i % 3)
                );
                order.addItem(item);

                // Set different statuses
                if (i % 3 === 1) {
                    order.confirm();
                } else if (i % 3 === 2) {
                    order.confirm();
                    order.ship();
                }

                await repository.addAsync(order);
            }
        });

        it('should find orders by customer ID', async () => {
            // Given
            const customerId = new CustomerId('customer-1');

            // When
            const result = await repository.findByCustomerIdAsync(customerId);

            // Then
            expect(result.isSuccess).toBe(true);
            expect(result.value.length).toBeGreaterThan(0);

            for (const order of result.value) {
                expect(order.customerId.equals(customerId)).toBe(true);
            }
        });

        it('should find orders by status', async () => {
            // Given
            const status = OrderStatus.CONFIRMED;

            // When
            const result = await repository.findByStatusAsync(status);

            // Then
            expect(result.isSuccess).toBe(true);
            expect(result.value.length).toBeGreaterThan(0);

            for (const order of result.value) {
                expect(order.status).toBe(status);
            }
        });

        it('should find orders by customer and status', async () => {
            // Given
            const customerId = new CustomerId('customer-1');
            const status = OrderStatus.PENDING;

            // When
            const result = await repository.findByCustomerAndStatusAsync(customerId, status);

            // Then
            expect(result.isSuccess).toBe(true);

            for (const order of result.value) {
                expect(order.customerId.equals(customerId)).toBe(true);
                expect(order.status).toBe(status);
            }
        });

        it('should find orders by date range', async () => {
            // Given
            const now = new Date();
            const oneHourAgo = new Date(now.getTime() - 60 * 60 * 1000);
            const oneHourFromNow = new Date(now.getTime() + 60 * 60 * 1000);

            // When
            const result = await repository.findByDateRangeAsync(oneHourAgo, oneHourFromNow);

            // Then
            expect(result.isSuccess).toBe(true);
            expect(result.value.length).toBe(10); // All orders should be in this range

            for (const order of result.value) {
                expect(order.createdAt.getTime()).toBeGreaterThanOrEqual(oneHourAgo.getTime());
                expect(order.createdAt.getTime()).toBeLessThanOrEqual(oneHourFromNow.getTime());
            }
        });

        it('should find orders by amount range', async () => {
            // Given
            const minAmount = 15;
            const maxAmount = 25;
            const currency = 'USD';

            // When
            const result = await repository.findByAmountRangeAsync(minAmount, maxAmount, currency);

            // Then
            expect(result.isSuccess).toBe(true);

            for (const order of result.value) {
                expect(order.totalAmount.amount).toBeGreaterThanOrEqual(minAmount);
                expect(order.totalAmount.amount).toBeLessThanOrEqual(maxAmount);
                expect(order.totalAmount.currency).toBe(currency);
            }
        });

        it('should get recent orders with limit', async () => {
            // Given
            const limit = 5;

            // When
            const result = await repository.findRecentOrdersAsync(limit);

            // Then
            expect(result.isSuccess).toBe(true);
            expect(result.value.length).toBeLessThanOrEqual(limit);

            // Verify orders are sorted by creation date (most recent first)
            for (let i = 1; i < result.value.length; i++) {
                expect(result.value[i - 1].createdAt.getTime()).toBeGreaterThanOrEqual(
                    result.value[i].createdAt.getTime()
                );
            }
        });

        it('should count orders by status', async () => {
            // Given
            const status = OrderStatus.PENDING;

            // When
            const result = await repository.countByStatusAsync(status);

            // Then
            expect(result.isSuccess).toBe(true);
            expect(result.value).toBeGreaterThan(0);
            expect(typeof result.value).toBe('number');
        });
    });

    describe('Should_HandleAdvancedOperations_When_PerformingComplexQueries', () => {
        beforeEach(async () => {
            // Set up more complex test data
            const customers = ['customer-A', 'customer-B'];

            for (let i = 0; i < 6; i++) {
                const customerId = new CustomerId(customers[i % 2]!);
                const order = Order.create(customerId);

                // Add different products
                const products = ['Laptop', 'Mouse', 'Keyboard'];
                const product = products[i % 3]!;
                const item = OrderItem.create(product, new Money(100 + i * 50, 'USD'), 1);
                order.addItem(item);

                if (i >= 3) {
                    order.confirm();
                }

                await repository.addAsync(order);
            }
        });

        it('should get customer statistics', async () => {
            // Given
            const customerId = new CustomerId('customer-A');

            // When
            const result = await repository.getCustomerStatisticsAsync(customerId);

            // Then
            expect(result.isSuccess).toBe(true);

            const stats = result.value;
            expect(stats.customerId).toBe('customer-A');
            expect(stats.totalOrders).toBe(3); // customer-A has 3 orders
            expect(stats.totalAmount).toBeGreaterThan(0);
            expect(stats.averageOrderValue).toBe(stats.totalAmount / stats.totalOrders);
            expect(stats.currency).toBe('USD');
            expect(stats.ordersByStatus).toBeDefined();
            expect(stats.mostOrderedProducts).toBeDefined();
        });

        it('should find orders with specific product', async () => {
            // Given
            const productName = 'Laptop';

            // When
            const result = await repository.findOrdersWithProductAsync(productName);

            // Then
            expect(result.isSuccess).toBe(true);

            for (const order of result.value) {
                expect(order.hasProduct(productName)).toBe(true);
            }
        });

        it('should get total sales for date range', async () => {
            // Given
            const now = new Date();
            const oneHourAgo = new Date(now.getTime() - 60 * 60 * 1000);
            const oneHourFromNow = new Date(now.getTime() + 60 * 60 * 1000);
            const currency = 'USD';

            // When
            const result = await repository.getTotalSalesAsync(oneHourAgo, oneHourFromNow, currency);

            // Then
            expect(result.isSuccess).toBe(true);
            expect(result.value).toBeGreaterThan(0);
            expect(typeof result.value).toBe('number');
        });

        it('should find stale orders', async () => {
            // Given
            const daysOld = 0; // Look for orders older than 0 days (none should exist)

            // When
            const result = await repository.findStaleOrdersAsync(daysOld);

            // Then
            expect(result.isSuccess).toBe(true);
            expect(result.value).toHaveLength(0); // No stale orders in fresh test data
        });

        it('should perform bulk operations', async () => {
            // Given
            const allOrders = repository.getAllOrders();
            const pendingOrders = allOrders.filter(o => o.status === OrderStatus.PENDING);
            const orderIds = pendingOrders.map(o => o.id);

            // When
            const result = await repository.bulkUpdateStatusAsync(orderIds, OrderStatus.CANCELLED);

            // Then
            expect(result.isSuccess).toBe(true);
            expect(result.value).toBe(orderIds.length);

            // Verify orders were updated
            for (const orderId of orderIds) {
                const orderResult = await repository.getByIdAsync(orderId);
                const order = orderResult.value.value;
                expect(order.status).toBe(OrderStatus.CANCELLED);
            }
        });

        it('should save multiple orders in transaction', async () => {
            // Given
            const customerId = new CustomerId('customer-batch');
            const orders: Order[] = [];

            for (let i = 0; i < 3; i++) {
                const order = Order.create(customerId);
                const item = OrderItem.create(`Product ${i}`, new Money(10 + i, 'USD'), 1);
                order.addItem(item);
                orders.push(order);
            }

            // When
            const result = await repository.saveAllAsync(orders);

            // Then
            expect(result.isSuccess).toBe(true);

            // Verify all orders were saved
            for (const order of orders) {
                const getResult = await repository.getByIdAsync(order.id);
                expect(getResult.isSuccess).toBe(true);
                expect(getResult.value.hasValue).toBe(true);
            }
        });
    });

    describe('Should_HandleErrorScenarios_When_OperationsFail', () => {
        it('should handle attempts to add duplicate orders', async () => {
            // Given
            const customerId = new CustomerId('customer-123');
            const order = Order.create(customerId);
            await repository.addAsync(order);

            // When - Try to add same order again
            const duplicateResult = await repository.addAsync(order);

            // Then
            expect(duplicateResult.isFailure).toBe(true);
            expect(duplicateResult.error.code).toBe('Repository.DuplicateId');
        });

        it('should handle attempts to update non-existent orders', async () => {
            // Given
            const customerId = new CustomerId('customer-123');
            const order = Order.create(customerId);

            // When - Try to update order that was never added
            const updateResult = await repository.updateAsync(order);

            // Then
            expect(updateResult.isFailure).toBe(true);
            expect(updateResult.error.code).toBe('Repository.NotFound');
        });

        it('should handle attempts to delete non-existent orders', async () => {
            // Given
            const nonExistentId = OrderId.generate();

            // When
            const deleteResult = await repository.deleteAsync(nonExistentId);

            // Then
            expect(deleteResult.isFailure).toBe(true);
            expect(deleteResult.error.code).toBe('Repository.NotFound');
        });

        it('should return empty results for queries with no matches', async () => {
            // Given
            const nonExistentCustomerId = new CustomerId('non-existent-customer');

            // When
            const result = await repository.findByCustomerIdAsync(nonExistentCustomerId);

            // Then
            expect(result.isSuccess).toBe(true);
            expect(result.value).toHaveLength(0);
        });
    });

    describe('Should_HandleCancellation_When_UsingAbortSignal', () => {
        it('should respect cancellation tokens in async operations', async () => {
            // Given
            const customerId = new CustomerId('customer-123');
            const order = Order.create(customerId);
            const abortController = new AbortController();

            // When - Cancel immediately
            abortController.abort();

            const addResult = await repository.addAsync(order, abortController.signal);

            // Then
            expect(addResult.isFailure).toBe(true);
            expect(addResult.error.code).toBe('Repository.OperationCancelled');
        });

        it('should handle cancellation during query operations', async () => {
            // Given
            const customerId = new CustomerId('customer-123');
            const abortController = new AbortController();

            // When - Cancel immediately
            abortController.abort();

            const findResult = await repository.findByCustomerIdAsync(customerId, abortController.signal);

            // Then
            expect(findResult.isFailure).toBe(true);
            expect(findResult.error.code).toBe('Repository.OperationCancelled');
        });
    });

    describe('Should_HandlePerformance_When_ProcessingManyOrders', () => {
        it('should handle large number of orders efficiently', async () => {
            // Given
            const customerId = new CustomerId('performance-customer');
            const orderCount = 1000;

            // When - Add many orders
            const startTime = performance.now();

            const addPromises: Promise<Result>[] = [];
            for (let i = 0; i < orderCount; i++) {
                const order = Order.create(customerId);
                const item = OrderItem.create(`Product ${i}`, new Money(10, 'USD'), 1);
                order.addItem(item);
                addPromises.push(repository.addAsync(order));
            }

            await Promise.all(addPromises);

            const addTime = performance.now() - startTime;

            // Then - Should complete within reasonable time
            expect(addTime).toBeLessThan(5000); // 5 seconds for 1000 orders

            // When - Query all orders for customer
            const queryStartTime = performance.now();
            const findResult = await repository.findByCustomerIdAsync(customerId);
            const queryTime = performance.now() - queryStartTime;

            // Then - Query should be fast
            expect(findResult.isSuccess).toBe(true);
            expect(findResult.value.length).toBe(orderCount);
            expect(queryTime).toBeLessThan(100); // Should complete within 100ms
        });

        it('should handle concurrent operations efficiently', async () => {
            // Given
            const customerIds = Array.from({ length: 10 }, (_, i) => new CustomerId(`customer-${i}`));

            // When - Perform concurrent operations
            const startTime = performance.now();

            const operations = customerIds.map(async (customerId) => {
                const order = Order.create(customerId);
                const item = OrderItem.create('Product', new Money(10, 'USD'), 1);
                order.addItem(item);

                await repository.addAsync(order);
                const getResult = await repository.getByIdAsync(order.id);
                return getResult;
            });

            const results = await Promise.all(operations);

            const endTime = performance.now();
            const duration = endTime - startTime;

            // Then - All operations should succeed
            for (const result of results) {
                expect(result.isSuccess).toBe(true);
                expect(result.value.hasValue).toBe(true);
            }

            expect(duration).toBeLessThan(1000); // Should complete within 1 second
        });
    });
});

File: golang/tests/integration/repository_scenarios_test.go
package integration

import (
	"context"
	"testing"
	"time"

	"github.com/universalddd/architecture-core-go/examples/quickstart"
	"github.com/universalddd/architecture-core-go/pkg/functional"
)

// TestRepositoryScenarios_Should_HandleCRUDOperations_When_Used
func TestRepositoryScenarios_Should_HandleCRUDOperations_When_Used(t *testing.T) {
	t.Run("Should_AddAndRetrieve_When_ValidAggregate", func(t *testing.T) {
		// Given: Repository and aggregate
		repo := quickstart.NewInMemoryOrderRepository()
		ctx := context.Background()

		order := quickstart.NewOrder("CUST-001", quickstart.NewMoney(100.00, "USD"))
		orderID := order.GetID()

		// When: Adding aggregate
		addResult := repo.AddAsync(ctx, order)

		// Then: Should succeed
		if addResult.IsFailure() {
			t.Fatalf("Add operation failed: %v", addResult.Error())
		}

		// When: Retrieving by ID
		getResult := repo.GetByIDAsync(ctx, orderID)

		// Then: Should return the aggregate
		if !getResult.HasValue() {
			t.Fatal("Should retrieve added aggregate")
		}

		retrievedOrder := getResult.Value()
		if retrievedOrder.GetID() != orderID {
			t.Error("Retrieved order should have same ID")
		}
	})

	t.Run("Should_UpdateAggregate_When_ModificationsExist", func(t *testing.T) {
		// Given: Repository with existing aggregate
		repo := quickstart.NewInMemoryOrderRepository()
		ctx := context.Background()

		order := quickstart.NewOrder("CUST-002", quickstart.NewMoney(200.00, "USD"))
		orderID := order.GetID()

		repo.AddAsync(ctx, order)

		// When: Modifying and updating aggregate
		order.ConfirmOrder()
		updateResult := repo.UpdateAsync(ctx, order)

		// Then: Should succeed
		if updateResult.IsFailure() {
			t.Fatalf("Update operation failed: %v", updateResult.Error())
		}

		// When: Retrieving updated aggregate
		getResult := repo.GetByIDAsync(ctx, orderID)
		updatedOrder := getResult.Value()

		// Then: Should reflect changes
		if updatedOrder.GetStatus() != quickstart.Confirmed {
			t.Error("Updated order should have Confirmed status")
		}
	})

	t.Run("Should_DeleteAggregate_When_IDProvided", func(t *testing.T) {
		// Given: Repository with existing aggregate
		repo := quickstart.NewInMemoryOrderRepository()
		ctx := context.Background()

		order := quickstart.NewOrder("CUST-003", quickstart.NewMoney(300.00, "USD"))
		orderID := order.GetID()

		repo.AddAsync(ctx, order)

		// When: Deleting aggregate
		deleteResult := repo.DeleteAsync(ctx, orderID)

		// Then: Should succeed
		if deleteResult.IsFailure() {
			t.Fatalf("Delete operation failed: %v", deleteResult.Error())
		}

		// When: Attempting to retrieve deleted aggregate
		getResult := repo.GetByIDAsync(ctx, orderID)

		// Then: Should not find it
		if getResult.HasValue() {
			t.Error("Should not retrieve deleted aggregate")
		}
	})

	t.Run("Should_CheckExistence_When_AggregateExistsOrNot", func(t *testing.T) {
		// Given: Repository with one aggregate
		repo := quickstart.NewInMemoryOrderRepository()
		ctx := context.Background()

		order := quickstart.NewOrder("CUST-004", quickstart.NewMoney(400.00, "USD"))
		orderID := order.GetID()

		repo.AddAsync(ctx, order)

		// When: Checking existence of existing aggregate
		existsResult := repo.ExistsAsync(ctx, orderID)

		// Then: Should return true
		if existsResult.IsFailure() || !existsResult.Value() {
			t.Error("Should return true for existing aggregate")
		}

		// When: Checking existence of non-existent aggregate
		notExistsResult := repo.ExistsAsync(ctx, "NON-EXISTENT")

		// Then: Should return false
		if notExistsResult.IsFailure() || notExistsResult.Value() {
			t.Error("Should return false for non-existent aggregate")
		}
	})

	t.Run("Should_HandleConcurrentOperations_When_MultipleClients", func(t *testing.T) {
		// Given: Repository and multiple goroutines
		repo := quickstart.NewInMemoryOrderRepository()
		ctx := context.Background()

		// When: Performing concurrent adds
		results := make(chan functional.Result, 10)

		for i := 0; i < 10; i++ {
			go func(index int) {
				customerID := "CUST-CONCURRENT-" + string(rune('A'+index))
				order := quickstart.NewOrder(customerID, quickstart.NewMoney(100.00, "USD"))
				result := repo.AddAsync(ctx, order)
				results <- result
			}(i)
		}

		// Then: All operations should succeed
		successCount := 0
		for i := 0; i < 10; i++ {
			result := <-results
			if result.IsSuccess() {
				successCount++
			} else {
				t.Errorf("Concurrent add failed: %v", result.Error())
			}
		}

		if successCount != 10 {
			t.Errorf("Expected 10 successful concurrent adds, got %d", successCount)
		}
	})

	t.Run("Should_RespectCancellation_When_ContextCancelled", func(t *testing.T) {
		// Given: Repository and cancelled context
		repo := quickstart.NewInMemoryOrderRepository()
		ctx, cancel := context.WithCancel(context.Background())
		cancel() // Cancel immediately

		order := quickstart.NewOrder("CUST-005", quickstart.NewMoney(500.00, "USD"))

		// When: Performing operation with cancelled context
		result := repo.AddAsync(ctx, order)

		// Then: Should handle cancellation gracefully
		// (Implementation may succeed quickly or return cancellation error)
		if result.IsFailure() {
			// Check if it's a cancellation-related error
			if result.Error().Category() == functional.Infrastructure {
				t.Log("Repository properly handled context cancellation")
			}
		}
	})

	t.Run("Should_RespectTimeout_When_ContextHasTimeout", func(t *testing.T) {
		// Given: Repository and context with timeout
		repo := quickstart.NewInMemoryOrderRepository()
		ctx, cancel := context.WithTimeout(context.Background(), 1*time.Millisecond)
		defer cancel()

		order := quickstart.NewOrder("CUST-006", quickstart.NewMoney(600.00, "USD"))

		// When: Performing operation with timeout context
		// Add slight delay to potentially trigger timeout
		time.Sleep(2 * time.Millisecond)
		result := repo.AddAsync(ctx, order)

		// Then: Should handle timeout appropriately
		// (May succeed if operation is fast, or return timeout error)
		if result.IsFailure() {
			if result.Error().Category() == functional.Infrastructure {
				t.Log("Repository properly handled context timeout")
			}
		}
	})

	t.Run("Should_PreventDuplicates_When_SameIDAdded", func(t *testing.T) {
		// Given: Repository with existing aggregate
		repo := quickstart.NewInMemoryOrderRepository()
		ctx := context.Background()

		order1 := quickstart.NewOrderWithID("DUPLICATE-ID", "CUST-007", quickstart.NewMoney(700.00, "USD"))
		order2 := quickstart.NewOrderWithID("DUPLICATE-ID", "CUST-008", quickstart.NewMoney(800.00, "USD"))

		repo.AddAsync(ctx, order1)

		// When: Adding duplicate ID
		duplicateResult := repo.AddAsync(ctx, order2)

		// Then: Should handle appropriately (fail or overwrite - implementation specific)
		if duplicateResult.IsFailure() {
			// Verify it's an appropriate error
			if duplicateResult.Error().Category() != functional.Domain {
				t.Error("Duplicate ID should return domain error")
			}
		} else {
			t.Log("Repository allows overwriting duplicates - acceptable behavior")
		}
	})

	t.Run("Should_HandleLargeOperations_When_ManyAggregates", func(t *testing.T) {
		// Given: Repository and many aggregates
		repo := quickstart.NewInMemoryOrderRepository()
		ctx := context.Background()

		const numOrders = 1000
		orderIDs := make([]string, numOrders)

		// When: Adding many aggregates
		for i := 0; i < numOrders; i++ {
			customerID := "BULK-CUST-" + string(rune('A'+(i%26)))
			order := quickstart.NewOrder(customerID, quickstart.NewMoney(float64(i+1), "USD"))
			orderIDs[i] = order.GetID()

			result := repo.AddAsync(ctx, order)
			if result.IsFailure() {
				t.Fatalf("Bulk add failed at index %d: %v", i, result.Error())
			}
		}

		// Then: All should be retrievable
		retrievedCount := 0
		for _, orderID := range orderIDs {
			getResult := repo.GetByIDAsync(ctx, orderID)
			if getResult.HasValue() {
				retrievedCount++
			}
		}

		if retrievedCount != numOrders {
			t.Errorf("Expected to retrieve %d orders, got %d", numOrders, retrievedCount)
		}
	})

	t.Run("Should_HandleOptimisticConcurrency_When_VersionConflicts", func(t *testing.T) {
		// Given: Repository with versioned aggregate
		repo := quickstart.NewInMemoryOrderRepository()
		ctx := context.Background()

		order := quickstart.NewOrder("CUST-009", quickstart.NewMoney(900.00, "USD"))
		orderID := order.GetID()

		repo.AddAsync(ctx, order)

		// When: Two clients modify same aggregate
		getResult1 := repo.GetByIDAsync(ctx, orderID)
		getResult2 := repo.GetByIDAsync(ctx, orderID)

		order1 := getResult1.Value()
		order2 := getResult2.Value()

		// Simulate modifications
		order1.ConfirmOrder()
		order2.ConfirmOrder()

		// When: First client updates
		update1Result := repo.UpdateAsync(ctx, order1)

		// Then: Should succeed
		if update1Result.IsFailure() {
			t.Fatalf("First update should succeed: %v", update1Result.Error())
		}

		// When: Second client updates (stale version)
		update2Result := repo.UpdateAsync(ctx, order2)

		// Then: Should handle concurrency conflict appropriately
		if update2Result.IsFailure() {
			if update2Result.Error().Category() != functional.Concurrency {
				t.Error("Concurrency conflict should return concurrency error")
			}
		} else {
			t.Log("Repository allows last-write-wins - acceptable for in-memory implementation")
		}
	})

	t.Run("Should_SupportCustomQueries_When_RepositoryExtended", func(t *testing.T) {
		// Given: Extended repository with custom query methods
		repo := quickstart.NewInMemoryOrderRepository()
		ctx := context.Background()

		// Add orders with different statuses
		pendingOrder := quickstart.NewOrder("CUST-010", quickstart.NewMoney(100.00, "USD"))
		confirmedOrder := quickstart.NewOrder("CUST-011", quickstart.NewMoney(200.00, "USD"))
		confirmedOrder.ConfirmOrder()

		repo.AddAsync(ctx, pendingOrder)
		repo.AddAsync(ctx, confirmedOrder)

		// When: Querying by status (if supported)
		pendingResult := repo.GetOrdersByStatusAsync(ctx, quickstart.Pending)
		confirmedResult := repo.GetOrdersByStatusAsync(ctx, quickstart.Confirmed)

		// Then: Should return appropriate orders
		if pendingResult.IsSuccess() {
			pendingOrders := pendingResult.Value()
			if len(pendingOrders) < 1 {
				t.Error("Should find at least one pending order")
			}
		}

		if confirmedResult.IsSuccess() {
			confirmedOrders := confirmedResult.Value()
			if len(confirmedOrders) < 1 {
				t.Error("Should find at least one confirmed order")
			}
		}
	})
}